README.md
# @cowtech/tslint-config

## END OF DEVELOPMENT NOTICE - This package has been discontinued

CowTech style TSLint config

https://github.com/cowtech/tslint-config

Starting from version **5.10.0**, it is now based on [standard](https://standardjs.com/) which will differ on only few specific cases.

## Usage

In your tslint.json configuration file, insert the following statement:

```typescript
// ...
"extends": ["@cowtech/tslint-config"]
// ...
```

## Contributing to @cowtech/tslint-config

- Check out the latest master to make sure the feature hasn't been implemented or the bug hasn't been fixed yet.
- Check out the issue tracker to make sure someone already hasn't requested it and/or contributed it.
- Fork the project.
- Start a feature/bugfix branch.
- Commit and push until you are happy with your contribution.
- Make sure to add tests for it. This is important so I don't break it in a future version unintentionally.

## Copyright

Copyright (C) 2017 and above Shogun <mailto:shogun@cowtech.it>.

Licensed under the MIT license, which can be found at https://choosealicense.com/licenses/mit.


CHANGELOG.md
### 2020-03-09 / 5.13.1

- chore: Added end of development notice.

### 2019-02-24 / 5.13.0

- chore: Updated dependencies.

### 2019-02-02 / 5.12.4

- fix: Always require curly braces.

### 2019-02-02 / 5.12.3

- chore: Use NPM for locking.
- feat: Always require curly braces.

### 2019-01-14 / 5.12.2

- fix: Relax typedef rule a little.

### 2019-01-14 / 5.12.1

- fix: Fixed typo.

### 2019-01-14 / 5.12.0

- feat: Added typedef rule.

### 2018-10-08 / 5.11.1

- fix: Updated dependencies.

### 2018-08-11 / 5.11.0

- feat: Updated dependencies.

### 2018-06-24 / 5.10.5

- fix: Use avoid rules for quotes.

### 2018-06-23 / 5.10.4

- fix: Satisfy peer dependency.

### 2018-06-18 / 5.10.3

- chore: Add prettier to dev dependencies.

### 2018-06-18 / 5.10.2

- chore: Update dependencies scheme.

### 2018-06-18 / 5.10.1

- fix: Use double quotes in JSX.

### 2018-06-18 / 5.10.0

- feat: Switch to StandardJS.
- feat: Updated license links and badges.

### 2018-01-18 / 5.9.2

- Update to TSLint 5.9.x. Align back with versioning.

### 2017-12-11 / 5.9.1

### 2017-11-30 / 5.9.0

### 2017-10-29 / 5.8.0

### 2017-09-08 / 5.7.1

### 2017-09-08 / 5.7.0

### 2017-08-14 / 5.6.5

### 2017-08-14 / 5.6.4

### 2017-07-19 / 5.6.3

- feat: Switched to single quotes to be closer to StandardJS.

### 2017-07-12 / 5.6.2

- fix: Minor bugfix.

### 2017-07-09 / 5.6.1

### 2017-07-09 / 5.6.0

### 2017-07-09 / 5.5.0

### 2017-07-09 / 5.4.5

- feat: Update to TSLint 5.5.x.

### 2017-06-28 / 5.4.4

### 2017-06-28 / 5.4.3

- feat: Moved to @cowtech organization on npm.
- feat: Added deprecation message.

### 2017-06-28 / 5.4.2

- feat: Added deprecation message.

### 2017-06-10 / 5.4.1

- fix: Minor fixes.

### 2017-06-10 / 5.4.0

- feat: Update to TSLint 5.4.x.

### 2017-06-09 / 5.3.2

- fix: Minor bugfix.

### 2017-05-25 / 5.3.1

- fix: Minor fix.

### 2017-05-25 / 5.3.0

- feat: Updated to TSLint 5.3.

### 2017-05-15 / 5.1.5

- feat: Disabled misleadign rule.

### 2017-05-03 / 5.1.4

- feat: Disabled misleadign rule.

### 2017-05-01 / 5.1.3

- feat: Disabled misleadign rule.

### 2017-05-01 / 5.1.2

- feat: Added missing folder in package.json files section.

### 2017-05-01 / 5.1.1

- fix: Minor fixes and React preset.

### 2017-05-01 / 5.1.0

- feat: Fixed README.


config/node.js
/*
 * This file is part of the tslint-config-cowtech npm package. Copyright (C) 2017 and above Shogun <shogun@cowtech.it>.
 * Licensed under the MIT license, which can be found at http://www.opensource.org/licenses/mit-license.php.
 */

const MAX_LINE_LENGTH = 160;
const MAX_COMPLEXITY = 25;
const MAGIC_NUMBERS = [
  -1, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, // Commonly used base numbers
  25, 28, 30, 50, 100, 250, 1000, 5000, 10000, 1E6, 1E9, 3600, // Commonly used date/time related numbers
  // Common HTTP status codes
  200, 201, 202, 204, 205, 206,
  300, 301, 302, 304, 307, 308,
  400, 401, 403, 404, 405, 406, 408, 409, 410, 415, 422, 429, 499,
  500, 501, 502, 504, 599
];

module.exports = {
  rules: {
    // TypeScript Specific
    "adjacent-overload-signatures": true,
    "ban-types": false,
    "member-access": false,
    "member-ordering": [
      true,
      "public-static-field", "protected-static-field", "private-static-field",
      "public-static-method", "protected-static-method", "private-static-method",
      "public-instance-field", "protected-instance-field", "private-instance-field",
      "constructor",
      "public-instance-method", "protected-instance-method", "private-instance-method"
    ],
    "no-any": false,
    "no-empty-interface": true,
    "no-import-side-effect": false,
    "no-inferrable-types": false,
    "no-internal-module": false,
    "no-magic-numbers": [true, ...MAGIC_NUMBERS],
    "no-namespace": false,
    "no-non-null-assertion": false,
    "no-reference": true,
    "no-unnecessary-type-assertion": false,
    "no-var-requires": false,
    "only-arrow-functions": false,
    "prefer-for-of": true,
    "promise-function-async": false,
    "typedef": true,
    "typedef-whitespace": [
      true,
      {"call-signature": "nospace", "index-signature": "nospace", "parameter": "nospace", "property-declaration": "nospace", "variable-declaration": "nospace"},
      {"call-signature": "space", "index-signature": "space", "parameter": "space", "property-declaration": "space", "variable-declaration": "space"}
    ],
    "unified-signatures": true,
    // Functionality
    "await-promise": true,
    "ban": false,
    "curly": false,
    "encoding": true,
    "forin": true,
    "import-blacklist": false,
    "label-position": true,
    "no-arg": true,
    "no-bitwise": true,
    "no-conditional-assignment": true,
    "no-console": false,
    "no-construct": true,
    "no-debugger": true,
    "no-duplicate-super": true,
    "no-duplicate-variable": [true, "check-parameters"],
    "no-empty": true,
    "no-eval": true,
    "no-floating-promises": true,
    "no-for-in-array": true,
    "no-inferred-empty-object-type": true,
    "no-invalid-template-strings": true,
    "no-invalid-this": false,
    "no-misused-new": true,
    "no-null-keyword": false,
    "no-object-literal-type-assertion": true,
    "no-shadowed-variable": true,
    "no-sparse-arrays": true,
    "no-string-literal": false,
    "no-string-throw": true,
    "no-switch-case-fall-through": false,
    "no-unbound-method": false,
    "no-unsafe-any": false,
    "no-unsafe-finally": true,
    "no-unused-expression": true,
    "no-unused-variable": false, // Enable this is one of -–noUnusedParameters or -–noUnusedLocals is NOT set
    "no-use-before-declare": true,
    "no-var-keyword": true,
    "no-void-expression": false,
    "radix": true,
    "restrict-plus-operands": true,
    "strict-boolean-expressions": false,
    "strict-type-predicates": false,
    "switch-default": false,
    "triple-equals": true,
    "typeof-compare": true,
    "use-default-type-parameter": true,
    "use-isnan": true,
    // Maintainability
    "cyclomatic-complexity": [true, MAX_COMPLEXITY],
    "eofline": true,
    "indent": [true, "spaces"],
    "linebreak-style": [true, "LF"],
    "max-classes-per-file": false,
    "max-file-line-count": false,
    "max-line-length": [true, MAX_LINE_LENGTH],
    "no-default-export": false,
    "no-mergeable-namespace": true,
    "no-require-imports": false,
    "object-literal-sort-keys": false,
    "prefer-const": true,
    "trailing-comma": [true, {"multiline": "never", "singleline": "never"}],
    // Style
    "align": [true, "parameters", "statements"],
    "array-type": [true, "generic"],
    "arrow-parens": [true, "ban-single-arg-parens"],
    "arrow-return-shorthand": [true, "multiline"],
    "binary-expression-operand-order": true,
    "callable-types": true,
    "class-name": true,
    "comment-format": [true, "check-space", "check-uppercase"],
    "completed-docs": false,
    "file-header": false,
    "import-spacing": true,
    "interface-name": false,
    "interface-over-type-literal": true,
    "jsdoc-format": true,
    "match-default-export-name": false,
    "newline-before-return": true,
    "new-parens": true,
    "no-angle-bracket-type-assertion": true,
    "no-boolean-literal-compare": true,
    "no-consecutive-blank-lines": true,
    "no-irregular-whitespace": true,
    "no-parameter-properties": true,
    "no-reference-import": true,
    "no-trailing-whitespace": true,
    "no-unnecessary-callback-wrapper": true,
    "no-unnecessary-initializer": true,
    "no-unnecessary-qualifier": false,
    "number-literal-format": true,
    "object-literal-key-quotes": [true, "consistent-as-needed"],
    "object-literal-shorthand": true,
    "one-line": [true, "check-catch", "check-finally", "check-else", "check-open-brace"],
    "one-variable-per-declaration": true,
    "ordered-imports": false,
    "prefer-conditional-expression": true,
    "prefer-function-over-method": false,
    "prefer-object-spread": true,
    "prefer-method-signature": true,
    "prefer-switch": ["true", {"min-cases": 3}],
    "prefer-template": [true, "allow-single-concat"],
    "quotemark": [true, "double", "jsx-double", "avoid-escape", "no-template"],
    "return-undefined": true,
    "semicolon": [true, "always"],
    "space-before-function-paren": [true, {"anonymous": "never", "named": "never", "asyncArrow": "always", "method": "never", "constructor": "never"}],
    "switch-final-break": [true, "always"],
    "variable-name": [true, "check-format", "allow-leading-underscore", "allow-pascal-case", "allow-snake-case"],
    "whitespace": [true, "check-decl", "check-operator", "check-module", "check-separator", "check-type", "check-typecast"]
  }
};
